src/relayserver/RegistrationManager.ts

```diff
@@ -126,15 +127,29 @@ export class RegistrationManager {
           await this._handleStakedEvent()
           break
         case 'HubUnauthorized':
+          this.delayedEvents.push({ block: eventData.returnValues.removalBlock.toString(), eventData })
           unregistered = true
-          receipts = receipts.concat(await this._handleHubUnauthorizedEvent(eventData, currentBlock))
           break
         case 'StakeUnlocked':
+          this.delayedEvents.push({ block: eventData.returnValues.withdrawBlock.toString(), eventData })
           unregistered = true
-          receipts = receipts.concat(await this._handleUnstakedEvent(eventData, currentBlock))
           break
       }
     }
+
+    // handle HubUnauthorized, StakeUnlocked only after their due time
+
+    for (const eventData of this._extractDuePendingEvents(currentBlock)) {
+      switch (eventData.event) {
+        case 'HubUnauthorized':
+          receipts = receipts.concat(await this._handleHubUnauthorizedEvent(eventData, currentBlock))
+          break
+        case 'StakeUnlocked':
```

Comment:
Please note that there is a 'StakeWithdrawn' event.
It seems to be way more reasonable to react to the stake being withdrawn.

Also, **pay attention** to the fact that 'penalizeRelayManager' does **not check**  if the stake was unlocked or not. It can be added but seems redundant as it is in the best interests of the relay operator to withdraw ASAP, especially if it might have committed a penalizeable action.